calc-example/calc/src/type_check.rs

```diff
@@ -72,14 +61,12 @@ impl CheckExpression<'_> {
     }
 
     fn find_function(&self, f: FunctionId) -> Option<Function> {
-        find_function(self.db, self.program, f)
+        self.program.find_function(self.db, f)
     }
 
     fn report_error(&self, span: Span, message: String) {
-        Diagnostics::push(
-            self.db,
-            Diagnostic::new(span.start(self.db), span.end(self.db), message),
-        );
+        let (start, end) = span.absolute_locations(self.db, self.anchor);
```

Comment:
Doesn't resolving the anchors here break the reuse? You're going to have to re-execute this function if the source code changes and you re-collect the Diagnostics to get the updated line numbers.I'd expect the Diagnostics to contain spans, and them to be resolved after the call to `accumulated`.

Reply:
It will break the re-use, but given that the function is producing errors, that seems fine to me. Usually only a small minority of the code is producing errors at any point in time, and it's likely being actively edited.


The reason that I prefer this route is that the diagnostics now represent data that can be consumed wiithout reading into the database. My general preference is that the inputs/outputs of the program don't contain any tracked structs. If they do, you run the risk of modifying an input and then reading a tracked struct from the previous revision. Although I guess that's not necessarily worse than reading any output from the past revision.

Reply:
I'm mostly convinced by the errors are probably being actively edited argument. It's a trade-off between more reuse in the face of lots of errors and everything being slightly slower because the spans are more complicated.

I wasn't imagining exposing the tracked structs in the program io, just that the thing accumulating the diags does the resolution rather than pushing the diags. I agree that exposing tracked structs is bad, I think that using an old tracked struct could cause panics or completely invalid data if it was removed or replaced in the intern table.